keps/sig-network/2438-dual-stack-apiserver/README.md

```diff
@@ -349,109 +412,32 @@ allow the upgrade to proceed smoothly.)
-should be no flapping; both old and new apiservers will agree at all
-times about what endpoints are available _in the cluster's primary IP
-family_, and thus will always agree on the contents of the
-single-stack `Endpoints` resource.
-
-For `EndpointSlice`, current versions of kube-apiserver create a
-single `EndpointSlice` object named "`kubernetes`", and assume that
-other apiservers will do the same; they don't look at any other
-`EndpointSlice` objects when reconciling. To create dual-stack
-EndpointSlices, a new apiserver would need to create two objects with
-different names. Even if new apiservers use the same name as the old
-apiserver does for the primary-IP-family slice, old apiservers would
-never see the slice it creates for the secondary IP family. This means
-there would be no flapping, but also means that a stale
-secondary-IP-family `EndpointSlice` could be left behind after a
-downgrade if the new apiserver exited uncleanly. (However, in this
-scenario the Service ought to end up as single-stack at the end as
-well, so the secondary-IP-Family `EndpointSlice` ought to be ignored
-anyway.)
-
-#### Service vs EndpointSlice Sync Issues
-
-Given that the Service is updated only on apiserver restart, but the
-EndpointSlice is updated periodically by each apiserver, then whenever
-there is a mix of single-stack and dual-stack apiservers, it is
-possible to see any combination of service and endpoint
-dual-stack-ness:
-
-  - Single-stack Service, single-stack EndpointSlice: in this case,
-    the service will work fine; everyone will just ignore the
-    secondary IP family.
-
-  - Single-stack Service, dual-stack EndpointSlices: kube-proxy will
-    ignore the secondary IP family endpoints, since it has no
-    secondary-family ClusterIP to match them up with. If other clients
-    try to use the secondary-IP-family EndpointSlice values directly,
-    they should succeed, except in the "stale EndpointSlice" case
-    discussed above.
-
-  - Dual-stack Service, single-stack EndpointSlice: kube-proxy will
-    create working mappings for the primary ClusterIP, and a "broken"
-    (ie, `-j REJECT`) mapping for the secondary ClusterIP. Clients
-    explicitly attempting to connect to the secondary ClusterIP will
-    fail, but clients attempting to connect to
-    `kubernetes.default.svc.cluster.local` should succeed, even if
-    they try the secondary ClusterIP first, because they should fall
-    back to the primary ClusterIP after the secondary one fails.
-
-  - Dual-stack Service, dual-stack EndpointSlices: works fine. Any
-    still-single-stack apiservers will not be advertising an IP in the
-    secondary family, so the secondary-family EndpointSlice may only
-    point to a subset of masters.
+Some of the details of upgrades and downgrades are discussed above under
+[Service and EndpointSlice Reconciling](#service-and-endpointslice-reconciling).
+
+The current (1.23) apiserver code will not clean up a
```

Comment:
Will older versions correctly handle the dual-stack lease in etcd?

Reply:
```
<<[UNRESOLVED endpoints-in-etcd ]>>
Need to figure out how dual-stack endpoints will be stored in etcd in
the `lease` mode, so that old apiservers won't get confused by them
(and, eg, try to create a single Endpoints object with both sets of
IPs).
<<[/UNRESOLVED]>>
```

(but probably "it writes it to an object with a different name, but we also keep writing to the legacy single-stack list for a few releases until it's safe to remove the legacy code")